src/lib.rs

```diff
@@ -130,33 +130,35 @@ impl Metadata {
 /// Capability to convert a key into an node index.
 pub trait Key {
     /// Convert the key `self` to a `usize` index.
-    fn find<M: Miniconf>(&self) -> core::option::Option<usize>;
+    fn find<const D: usize, M: Miniconf<D>>(&self) -> core::option::Option<usize>;
 }
 
-macro_rules! key_integer {
-    ($($ty:ident)+) => {
-        $(
-            impl Key for $ty {
-                #[inline]
-                fn find<M>(&self) -> core::option::Option<usize> {
-                    Some(*self as _)
-                }
-            }
-        )+
+impl Key for usize {
+    #[inline]
+    fn find<const D: usize, M>(&self) -> core::option::Option<usize> {
+        Some(*self)
     }
 }
 
-key_integer!(usize);
-
 impl Key for &str {
     #[inline]
-    fn find<M: Miniconf>(&self) -> core::option::Option<usize> {
+    fn find<const D: usize, M: Miniconf<D>>(&self) -> core::option::Option<usize> {
         M::name_to_index(self)
     }
 }
 
 /// Trait exposing serialization/deserialization of nodes by keys/paths and traversal.
-pub trait Miniconf {
+///
+/// The parameter `Y` is a lower bound for the recursion depth. It's only non-unity in
+/// `array` and `option` as they need a configurable recursion depth at compile time
+/// and can't be marked up with inner attributes.
+/// The paramter `Y` deviates from the key tree depth significantly:
+/// * `Option` consume a layer of Miniconf recursion, but don't add to key tree height
+/// * A struct that has `Miniconf` derived for it will only implement `Miniconf<1>`
+///   but may perform arbitrarily recursion and also have arbitrarily long keys.
+///   `1` is then the minimum recursion depth for a (non-newtype) struct. No other
+///   value is needed or valid.
+pub trait Miniconf<const Y: usize> {
```

Comment:
I think we could make this default to 1, since that's the "normal" operational mode.
```suggestion
pub trait Miniconf<const Y: usize = 1> {
```

Reply:
Ah. Right! Even better